HumanCompatibleAI/KataGo-custom
# Merge upstream changes from lightvector/KataGo

Merge changes from upstream up to Nov 28 2022 in https://github.com/lightvector/KataGo/commit/3c42da50ec0e1dd7192ea384e9c643981b946d5e

Same as https://github.com/HumanCompatibleAI/KataGo-custom/pull/69 but with commit history tidied via squash rebase. I was only able to tidy the last part of it -- attempting to tidy earlier parts resulted in a bunch of merge conflicts. Could be done, but not automatically, as Git doesn't record the resolution of merge conflicts.

Note we will want to merge (not squash merge) this PR.

---

Notes from @tomtseng:

Notable behavior changes:
* Changes to LCB to give stronger performance at low visit counts
* Changes to FPU calculation (this can be turned off with the config param `fpuParentWeightByVisitedPolicy = false`)

I think the config changes in order to return close to v1.10.0 behavior (i.e., before this PR) are
```
useGraphSearch = false
valueWeightExponent = 0.5
subtreeValueBiasFactor = 0.35
subtreeValueBiasWeightExponent = 0.8
fpuParentWeightByVisitedPolicy = false
```

## Testing

Two tests to sanity check that the merge PRs won’t cause major issues for us:

1. Run a training run with victim visits = 1, check that we still find the passing attack. Turn gatekeeping on too to check that gatekeeper still works.
2. Run GTP games, with one player being pass-adv/cyclic-adv running on the merge PR and the other player being a victim running on `stable`. Check the win rate is similar to the same setup but with adv running on `stable` too

Results here:  [experiment page](https://www.notion.so/chaiberkeley/Nov-2022-merge-PR-v1-12-4-merge-PR-testing-af0cc4819eb043fe8451633c25fcc770)
